### DISCUSSION TITLE
Can i reduce priority for external scanner?

### DISCUSSION THREAD

**Discussion by pleshevskiy** [tree-sitter/tree-sitter#1975] [Q&A] [answered]
I created external scanner to match unquoted string, but I need also keywords and tree-sitter always seems to start with my external scanner.

```
keyword: value
```
^ I want mark this as a keyword
How can I reduce priority for external scanner?

```javascript
module.exports = grammar({
  name: "test",

  externals: ($) => [$._identifier],

  rules: {
    source_file: ($) => seq($.identifier, ':', /[a-z]+/),

    identifier: ($) => choice($.keyword, $._identifier),

    keyword: ($) => choice("keyword"),
  },
});
```

**Comment by ahelwer** [comment, marked as answer]
You can't. If the external scanner has a token that is valid/possible at that position it will always be called first. There is probably some other way of doing what you want.

**Reply by pleshevskiy** [reply]
What ways are there? I tried looking at `word`, `conflicts`, `precedents`, but I don't seem to understand how they work. I would be glad if you could give me more information.

**Reply by ahelwer** [reply]
You can encode absolutely arbitrary logic in your external scanner. Why not just return false if you see the string `keyword`? Or just move `keyword` to be a token in your external scanner. Or come up with some way of matching an unquoted string in your grammar instead of using the external scanner.
